server/gateway_test.go

```diff
@@ -432,6 +433,183 @@ func TestGatewayIgnoreSelfReference(t *testing.T) {
 	}
 }
 
+func TestGatewayHeaderInfo(t *testing.T) {
+	o := testDefaultOptionsForGateway("A")
+	s := runGatewayServer(o)
+	defer s.Shutdown()
+
+	gwconn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", o.Gateway.Host, o.Gateway.Port))
+	if err != nil {
+		t.Fatalf("Error dialing server: %v\n", err)
+	}
+	defer gwconn.Close()
+	client := bufio.NewReaderSize(gwconn, maxBufSize)
+	l, err := client.ReadString('\n')
+	if err != nil {
+		t.Fatalf("Error receiving info from server: %v\n", err)
+	}
+	var info serverInfo
+	if err = json.Unmarshal([]byte(l[5:]), &info); err != nil {
+		t.Fatalf("Could not parse INFO json: %v\n", err)
+	}
+	if !info.Headers {
+		t.Fatalf("Expected by default for header support to be enabled")
+	}
+
+	s.Shutdown()
+	gwconn.Close()
+
+	// Now turn headers off.
+	o.NoHeaderSupport = true
+	s = runGatewayServer(o)
+	defer s.Shutdown()
+
+	gwconn, err = net.Dial("tcp", fmt.Sprintf("%s:%d", o.Gateway.Host, o.Gateway.Port))
+	if err != nil {
+		t.Fatalf("Error dialing server: %v\n", err)
+	}
+	defer gwconn.Close()
+	client = bufio.NewReaderSize(gwconn, maxBufSize)
+	l, err = client.ReadString('\n')
+	if err != nil {
+		t.Fatalf("Error receiving info from server: %v\n", err)
+	}
+	if err = json.Unmarshal([]byte(l[5:]), &info); err != nil {
+		t.Fatalf("Could not parse INFO json: %v\n", err)
+	}
+	if info.Headers {
+		t.Fatalf("Expected header support to be disabled")
+	}
+}
+
+func TestGatewayHeaderSupport(t *testing.T) {
+	o2 := testDefaultOptionsForGateway("B")
+	o2.Gateway.ConnectRetries = 0
+	s2 := runGatewayServer(o2)
+	defer s2.Shutdown()
+
+	o1 := testGatewayOptionsFromToWithServers(t, "A", "B", s2)
+	s1 := runGatewayServer(o1)
+	defer s1.Shutdown()
+
+	// s1 should have an outbound gateway to s2.
+	waitForOutboundGateways(t, s1, 1, time.Second)
+	// s2 should have an inbound gateway
+	waitForInboundGateways(t, s2, 1, time.Second)
+	// and an outbound too
+	waitForOutboundGateways(t, s2, 1, time.Second)
+
+	c, cr, _ := newClientForServer(s1)
+	defer c.close()
+
+	connect := "CONNECT {\"headers\":true}"
+	subOp := "SUB foo 1"
+	pingOp := "PING\r\n"
+	cmd := strings.Join([]string{connect, subOp, pingOp}, "\r\n")
```

Comment:
Since you use `\r\n` here, do you need it in pingOp?

Reply:
Yes since join does not put it on the end and we need that to have the parser finalize and send back a PONG.